Load numeric initial data with iterable actions

nilsvu opened the pull request:
## Proposed changes

Switch the numeric initial data importing from custom code in `DgElementArray` to iterable actions.

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Load numeric initial data with iterable actions

nilsvu commented on src/IO/Importers/VolumeDataReaderActions.hpp:
```diff
@@ -83,115 +88,130 @@ namespace ThreadedActions {
-      tmpl::for_each<FieldTagsList>([
-        &all_tensor_data, &volume_file, &observation_id
-      ](auto field_tag_v) noexcept {
+    // Open the volume data file
```
Code between here and line 206 is unchanged. It only shows up in this diff because the indentation level is reduced by two spaces.

geoffrey4444 commented on src/Evolution/Protocols.hpp:
```diff
@@ -28,4 +28,19 @@ struct NumericInitialData {
+
+// FIXME: Is there a way to avoid adding this lazy metafunction?
+template <typename Type, typename = std::void_t<>>
+struct get_import_fields_impl {
```
Is it possible to revise your implementation to avoid using a protocol?

nilsvu replied:
The protocol is still needed to check if we're using numeric initial data, an analytic solution or analytic data. But I simplified it a lot so the choice _which_ fields are imported from the data file is just made in the Metavariables. Once we need the flexibility to choose these fields, we can revisit how to do that (perhaps even at runtime).

I pushed these changes as a fixup.

geoffrey4444 commented on src/IO/Importers/ElementActions.hpp:
```diff
@@ -53,5 +61,111 @@ struct RegisterWithVolumeDataReader {
+    auto& local_reader_component =
+        *Parallel::get_parallel_component<
+             importers::VolumeDataReader<Metavariables>>(cache)
+             .ckLocalBranch();
```
I half remember in our discussion yesterday that we should avoid calling .ckLocalBranch? So I understand why it's ok, could you please explain how this line works and remind me what the concern about .ckLocalBranch was?

nilsvu replied:
The `VolumeDataReader` nodegroup component has one chare on every node, and here I just retrieve that chare directly.

wthrowe replied:
@geoffrey4444 The concern was that the element would call an action synchronously on the nodegroup and then that would call an action synchronously on the element, which results in two actions running on the element at once.  As long as the second of those calls is asynchronous (i.e., does not use `ckLocalBranch`) the first action on the element will finish before the second one starts.  The call to the nodegroup (here) can still be done synchronously.

nilsdeppe replied:
While @wthrowe is right, the `receive_data` calls may be automatically inlined and so are synchronous. This is not true for `simple_action`, and so to be safe we need to make this an async call through Charm++

wthrowe replied:
Is it safe to call `receive_data` from a nodegroup at all then?  The asynchronous nodegroup action could be scheduled by Charm++ to run at the same time as an element action on a different thread, couldn't it?

nilsdeppe replied:
Sorry, I'm not sure I exactly understand the scenario you mean. If I understand correctly, you're worried about the nodegroup calling `receive_data` while the element is being worked on on another core? So, say the nodegroup action is being run on core 2, and it invokes `receive_data` on a element that is on core 4 (same node, different core). This call, as far as I understand Charm++, would not be inlined because `core 4` is treated as non-local for array chares, but local for nodegroups. That is, it is the callee that determines whether or not it is local, not the caller, I think.

If I've completely misunderstood the discussion, please let me know :)

wthrowe replied:
Ah, I see. That makes sense.

nilsvu replied:
I removed the `ckLocalBranch` in a fixup commit

geoffrey4444 commented on src/IO/Importers/ElementActions.hpp:
```diff
@@ -53,5 +61,111 @@ struct RegisterWithVolumeDataReader {
+            inboxes);
+    // Using `0` for the temporal ID since we only read the volume data once, so
+    // there's no need to keep track of the temporal ID.
+    const auto received_data = inbox.find(0_st);
```
Where is 0_st defined?

nilsvu replied:
Utilities/Literals.hpp

geoffrey4444 commented on tests/Unit/IO/Importers/Test_VolumeDataReaderActions.cpp:
```diff
@@ -120,14 +109,23 @@ SPECTRE_TEST_CASE("Unit.IO.Importers.VolumeDataReaderActions", "[Unit][IO]") {
                                               {1, {{{1, 0}, {2, 3}}}},
                                               {1, {{{1, 0}, {5, 4}}}},
                                               {0, {{{1, 0}, {1, 0}}}}};
+  std::unordered_map<ElementId<2>,
```
Please explain so I understand what these changes are doing?

nilsvu replied:
In this test I didn't move the read-in data into the DataBox, but had the sample data in the DataBox and compared both in the callback action. Now I begin with an empty DataBox, have the new `ReceiveVolumeData` action populate it with the read-in data and compare it to the sample data that lives outside (in the test case).

geoffrey4444 commented on src/IO/Importers/VolumeDataReaderActions.hpp:
```diff
@@ -83,115 +88,130 @@ namespace ThreadedActions {
+    // data files that will be read in at compile-time to initialize a flag in
+    // the DataBox for each of them.
+    const auto& has_read_volume_data = db::get<Tags::HasReadVolumeData>(box);
+    const auto volume_data_id = pretty_type::get_name<ImporterOptionsGroup>();
```
I'd like to double-check I understand what this is doing. Can you please give me an example of what ImporterOptionsGroup's type is, the corresponding value of volume_data_id, and how the next line checks to make sure the data has been read? I would have thought ImporterOptionsGroup contained all of the options from the group.

I guess my real question is why not just set a bool here?

nilsvu replied:
The importers dev guide (in this PR) has an example at the bottom. The `ImporterOptionsGroup` is the location in the input file from where we read the filename, subfile and observation ID, so it identifies the file to read in. The `volume_data_id` is just a string representation of the class name, so again, it identifies the file to read in. This ID is placed in the DataBox to say "we have read this file", so subsequent calls don't read it again. To make this a bool we would have to collect all possible option groups (i.e. files) that will be read in, so we can initialize a bool in the DataBox for each of the files.

geoffrey4444 reviewed (changes requested):
I have a few questions about how this works, but looks reasonable as far as I can understand. 

One big question I have: before I planned to reuse "SetData"...but now, it looks like I'll want to write my own ReceiveFunctionOfTimeData action...is that right? Along the same lines, will I still be able to reuse RegisterWithVolumeDataReader for element registration (i.e., can we just register elements once, so we don't have to register elements once for reading volume data and once for reading FunctionOfTime data?)?

nilsvu commented:
@geoffrey4444 Yes I think you should write a `ReceiveFunctionOfTimeData` action. In principle you could reuse the `VolumeDataReader` node group component including its registered elements (that part of the code didn't change). You may want to rename it though.. I'll have to think about that.

nilsvu commented:
@geoffrey4444 Yes I suppose we can rename `VolumeDataReader` to `ElementDataReader` and use it to load more than just volume data.

geoffrey4444 commented:
> @geoffrey4444 Yes I suppose we can rename `VolumeDataReader` to `ElementDataReader` and use it to load more than just volume data.

Let's do that in this PR!

nilsvu committed:
Rename VolumeDataReader to ElementDataReader

wthrowe commented on src/Evolution/Executables/GeneralizedHarmonic/EvolveGeneralizedHarmonic.hpp:
```diff
@@ -358,13 +358,24 @@ struct EvolutionMetavars {
               Parallel::PhaseActions<Phase, Phase::Initialization,
                                      initialization_actions>,
-              tmpl::conditional_t<
+              std::conditional_t<
```
The `tmpl` version is preferred.  It is more efficient* than the one in `std` because it does not require additional template instantiations for each use.

\*I don't think STL implementations are forbidden from using the version in `tmpl`, but in practice they do not.

nilsvu replied:
right, this was left-over from experimenting with short-circuiting

nilsvu committed:
Move importer actions into separate files

nilsvu commented:
I pushed a commit that renames `VolumeDataReader` -> `ElementDataReader` and another that moves the actions into separate files.

geoffrey4444 commented:
I'm happy for you to squash after others have taken a look

wthrowe reviewed (changes requested):
Squash directly.

nilsdeppe commented on src/IO/Importers/Tags.hpp:
```diff
@@ -170,6 +173,25 @@ struct RegisteredElements : db::SimpleTag {
 };
 
+/// Indicates which volume data files have already been read.
+struct HasReadVolumeData : db::SimpleTag {
```
What about `VolumeDataAlreadyRead`? I think the fun English `read` and `read` (present and past) being the same made me fumble when I read this the first time

nilsvu replied:
renamed

nilsdeppe reviewed (commented):
I just have the one comment. I'm fine with you squashing once others are happy

kidder commented on src/IO/Importers/ElementActions.hpp:
```diff
@@ -53,5 +61,111 @@ struct RegisterWithVolumeDataReader {
+ * \see Dev guide on \ref dev_guide_importing
+ */
+template <typename ImporterOptionsGroup, typename FieldTagsList>
+struct ReadVolumeData {
```
why does this action exist?  why can't the `VolumeDataReader` invoke `ReadAllVolumeDataAndDistribute` in its own phase action list?

nilsvu replied:
Synchronizing PDALs between parallel components doesn't work very well. We could construct the PDAL for the `ElementDataReader` nodegroup in the metavariables, including the `InitializeElementDataReader` action for the init phase. Not sure if that's preferable.

Currently, parallel components either hard-code specific phases like `Initialization` and `Register` (e.g. observers, and interpolators) or the full PDAL is constructed in the metavariables (e.g. DgElementArray). Switching a singleton or nodegroup to the construct-pdal-in-metavariables approach means that the parallel component code does pretty much nothing but specify that it is a singleton or nodegroup. Array components at least have their allocation code left.

Edit: I tried to implement a version where we would construct the PDAL for the `ElementDataReader` in the metavariables. I think it could be a nice way to go usability-wise, if not for this hard problem: For registration, the elements need to invoke a simple action on the reader component, so they need to instantiate the reader component template. But the reader component sends data to the elements by invoking `receive_data`, so it needs to know the element array template instantiation. Currently, the element array template instantiation is passed to the simple action that's invoked on the reader component, and the reader component is only templated on `Metavariables` so it can be instantiated from every action. But changing `ReadVolumeData` to an iterable action makes the template instantiations circularly depend on each other. Does this make sense? Can anyone think of a solution here?

kidder replied:
okay, thank you for looking into this; it is not worth holding up this PR

nilsvu committed:
Remove a ckLocalBranch

geoffrey4444 commented:
I'm fine for you to squash

geoffrey4444 commented:
You have a conflict...please rebase?

nilsvu commented:
Rebased to fix the conflict. @wthrowe @nilsdeppe could you check the fixup (just making sure it's correct to remove the `ckLocalBranch`). I also replied to @kidder's comment.

nilsvu commented:
@nilsdeppe @wthrowe @kidder could you take a look at this PR and check the fixup?

wthrowe commented:
Looks fine.

nilsdeppe reviewed (commented):
I'm happy too

kidder commented:
you can squash

nilsvu commented:
Rebased on develop, and renamed the last commit instead of squashing it to avoid dealing with the refactors in the commit history. Thanks for reviewing everyone!

kidder commented:
@geoffrey4444 please look at

nilsdeppe commented:
Restarted CI because it failed with a random number test
